papis/isbn.py

```diff
@@ -11,24 +17,256 @@
+
+def googlebooks_isbn_search(query: str) -> str:
+    # This assumes the query to be an ISBN
+    query = strip_isbnlike(query)
+    base_api_link = "https://www.googleapis.com/books/v1/volumes?q=isbn:"
+    data = json_request(requests.utils.requote_uri(base_api_link + query))
+    if not data:
+        return ""
+    try:
+        items = data["items"]
+    except KeyError:
+        return ""
+    # Return the first ISBN that was provided by the API
+    for i in items:
+        for d in i.get("volumeInfo", {}).get("industryIdentifiers", {}):
+            # isbnlib also doesn't consider ISBN-10 results
+            if d.get("type") == "ISBN_13":
+                return str(d["identifier"])
+    return ""
+
+
+def isbn_from_words_fallback(query: str) -> str:
+    search_provider_url = "http://www.google.com/search?q="
+    search_url = (
+        search_provider_url
+        + "ISBN+"
+        + requests.utils.requote_uri(query.replace(" ", "+"))
+    )
+    r = requests.get(
+        search_url,
+        headers={
+            "user-agent": USER_AGENT,
+            "Content-Type": 'text/plain; charset="UTF-8"',
+            "Content-Transfer-Encoding": "Quoted-Printable",
+        },
+    )
+    regex = re.compile(
+        r"97[89]-?[0-9]{10}|" r"97[89]-[-0-9]{13}|" r"[-0-9]{9,15}[0-9xX]",
+        re.M,
+    )
+    potential_isbns = regex.findall(r.text)
+    isbn = ""
+    for i in potential_isbns:
+        if not notisbn_fallback(i):
+            isbn = googlebooks_isbn_search(i)
+            if len(isbn) > 0:
+                break
+    return isbn
+
+
+def meta_goob(isbn: str) -> dict[str, Any] | None:
+    url = (
+        "https://www.googleapis.com/books/v1/volumes?q={isbn_query}&fields="
+        "items/volumeInfo(title,subtitle,authors,publisher,publishedDate,"
+        "language,industryIdentifiers,description,imageLinks)&maxResults=1"
+    )
+    data = json_request(url.format(isbn_query="isbn:" + isbn))
+    # Apparently this sometimes works when the previous request doesn't.
+    # See: https://github.com/xlcnd/isbnlib/issues/119
+    if not data:
+        data = json_request(url.format(isbn_query=isbn))
+    if not data:
+        return None
+    try:
+        volumeinfo = data["items"][0]["volumeInfo"]
+    except KeyError:
+        return None
+    # Check that ISBNs match
+    ids = repr(volumeinfo.get("industryIdentifiers", ""))
+    if "ISBN_13" in ids and isbn not in ids:
+        raise ValueError("ISBN mismatch")
+    meta: dict[str, Any] = {}
+    meta["ISBN-13"] = isbn
+    title = volumeinfo.get("title", "").replace(" :", ":")
+    subtitle = volumeinfo.get("subtitle", "")
+    meta["Title"] = format_title(title, subtitle)
+    meta["Authors"] = volumeinfo.get("authors", [""])
+    meta["Publisher"] = volumeinfo.get("publisher", "").strip('"')
+    if "publishedDate" in volumeinfo and len(volumeinfo["publishedDate"]) >= 4:
+        meta["Year"] = volumeinfo["publishedDate"][0:4]
+    else:
+        meta["Year"] = ""
```

Comment:
Would it be better to just not add these fields if they don't exist in `volumeinfo`? What does `isbnlib` do?